.github/workflows/tests.yml

```diff
@@ -37,49 +37,93 @@ jobs:
         uses: actions/setup-python@v2
         with:
           python-version: ${{ matrix.python-version }}
+
       - name: Upgrade pip
         run: |
           pip install --upgrade --user pip
+
       - name: Get pip cache dir
         id: pip-cache
         run: |
           echo "::set-output name=dir::$(pip cache dir)"
+
       - name: pip cache
         uses: actions/cache@v2
         with:
           path: ${{ steps.pip-cache.outputs.dir }}
           key: ${{ runner.os }}-pip-py${{ matrix.python-version }}-${{ hashFiles('**/requirements.txt') }}
           restore-keys: |
             ${{ runner.os }}-pip-py${{ matrix.python-version }}-
-      - name: Install CPU Torch
-        if: runner.os != 'macos'
-        run: |
-          pip install torch==1.8.1+cpu torchvision==0.9.1+cpu torchcsprng==0.2.1+cpu -f https://download.pytorch.org/whl/torch_stable.html
-        shell: bash
-      - name: Install dependencies
-        run: |
-          hash -r
-          pip install -r requirements.dev.txt -f https://download.pytorch.org/whl/torch_stable.html
-          pip install -e .
-        shell: bash
+
       - id: file_changes # get changed files.
         uses: jitterbit/get-changed-files@v1
         with:
           format: space-delimited
 
-      - uses: pre-commit/action@v2.0.0
+      - uses: pre-commit/action@v2.0.3
         with:
           extra_args: --files ${{ steps.file_changes.outputs.all }} # apply only to changed files.
 
       - name: Check docstring coverage
         run: |
+          pip install interrogate
           interrogate -mvp src/ -f 80
 
       - name: Scan for security issues
         run: |
+          pip install bandit safety
           bandit -r src
           safety check
 
+
+  python-tests:
+    needs: [python-linting]
+    strategy:
+      max-parallel: 2
+      matrix:
+        os: [ubuntu-latest, macos-latest]
```

Comment:
Q: Can't we have only ubuntu-latest here also?

Reply:
For the normal tests, we would have only ubuntu?

Reply:
Ahh...those are the actual tests.
I read `needs: [python-linting]` as `those are the python linting tests`